WalletSwapCoin: Split into two classes + add trait

GeneFerneau opened the pull request:
Convert `WalletSwapCoin` into `Incoming/OutgoingSwapCoin`, and add a `WalletSwapCoin` trait for common functionality

chris-belcher commented:
Thanks for the PR(!)

chris-belcher commented on src/contracts.rs:
```diff
@@ -44,6 +46,9 @@ pub trait SwapCoin {
     fn get_multisig_redeemscript(&self) -> Script;
     fn get_contract_tx(&self) -> Transaction;
     fn get_contract_redeemscript(&self) -> Script;
+    fn get_other_privkey(&self) -> Option<&SecretKey>;
+    fn get_contract_privkey(&self) -> Option<&SecretKey>;
+    fn get_hash_preimage(&self) -> Option<&[u8; 32]>;
```
These functions should not be part of the trait SwapCoin.

Because WatchOnlySwapCoins will never have access to these, and our code ends up trying to get the hash preimage or contract privkey from a WatchOnlySwapCoin then the compiler should raise an error. I got the idea from this section https://doc.rust-lang.org/book/ch17-03-oo-design-patterns.html#encoding-states-and-behavior-as-types

My other comments right now will explain how the code can be modified to never require these three functions in the SwapCoin trait

GeneFerneau replied:
I removed these functions from `SwapCoin`. Added a couple other functions to get the contract type, and whether contract information is known (`other_privkey` for `IncomingSwapCoin`, `hash_preimage` for `Incoming/OutgoingSwapCoin`), and the type string.

chris-belcher commented on src/wallet_sync.rs:
```diff
@@ -91,127 +93,212 @@ const WATCH_ONLY_SWAPCOIN_LABEL: &str = "watchonly_swapcoin_label";
+pub struct OutgoingSwapCoin {
+    pub my_privkey: SecretKey,
+    pub other_pubkey: PublicKey,
+    pub other_privkey: Option<SecretKey>,
```
OutgoingSwapCoin will *never* have the `other_privkey`. That other privkey can only ever be given to us via private key handover (see `handle_private_key_handover`), which only ever happens for IncomingSwapCoins.

GeneFerneau replied:
Removed

chris-belcher commented on src/wallet_sync.rs:
```diff
@@ -91,127 +93,212 @@ const WATCH_ONLY_SWAPCOIN_LABEL: &str = "watchonly_swapcoin_label";
-            )
+
+        for (multisig_redeemscript, swapcoin) in &self.incoming_swap_coins {
+            Self::print_script_and_coin(multisig_redeemscript, swapcoin);
+        }
+        for (multisig_redeemscript, swapcoin) in &self.outgoing_swap_coins {
+            Self::print_script_and_coin(multisig_redeemscript, swapcoin);
         }
-        println!("swapcoin count = {}", self.swap_coins.len());
+        println!(
+            "swapcoin count = {}",
+            self.incoming_swap_coins.len() + self.outgoing_swap_coins.len()
+        );
+    }
+
+    fn print_script_and_coin(script: &Script, coin: &dyn SwapCoin) {
+        let contract_tx = coin.get_contract_tx();
+        println!(
+            "{} {}:{} {}",
+            Address::p2wsh(script, NETWORK),
+            contract_tx.input[0].previous_output.txid,
+            contract_tx.input[0].previous_output.vout,
+            if coin.get_other_privkey().is_some() {
+                "  known"
+            } else {
+                "unknown"
+            }
+        )
```
We can avoid using SwapCoin.get_other_privkey() here if we retain knowledge of the type (either IncomingSwapCoin or OutgoingSwapCoin). OutgoingSwapCoin will never have the field `other_privkey`. Probably this function `print_script_and_coin` should only take IncomingSwapCoins.

GeneFerneau replied:
> Probably this function print_script_and_coin should only take IncomingSwapCoins.

I think it's important to print all the coins in this context, and added an `is_known` function to `SwapCoin`. For `IncomingSwapCoin` is tests the presence of `hash_preimage` or `other_privkey`, and for `OutgoingSwapCoin` only tests for `hash_preimage`.

chris-belcher commented on src/main.rs:
```diff
@@ -191,17 +191,18 @@ fn display_wallet_balance(wallet_file_name: &PathBuf, long_form: Option<bool>) {
                     compressed: true,
                     key: bitcoin::secp256k1::PublicKey::from_secret_key(
                         &secp,
-                        &swapcoin.contract_privkey,
+                        swapcoin.get_contract_privkey().unwrap(),
                     ),
                 };
                 let type_string = if contract_pubkey
-                    == read_hashlock_pubkey_from_contract(&swapcoin.contract_redeemscript).unwrap()
+                    == read_hashlock_pubkey_from_contract(&swapcoin.get_contract_redeemscript())
+                        .unwrap()
                 {
                     "hashlock"
                 } else {
                     assert_eq!(
                         contract_pubkey,
-                        read_timelock_pubkey_from_contract(&swapcoin.contract_redeemscript)
+                        read_timelock_pubkey_from_contract(&swapcoin.get_contract_redeemscript())
                             .unwrap()
                     );
                     "timelock"
```
As "timelock" and "hashlock" exactly match the OutgoingSwapCoin and IncomingSwapCoin types, we no longer need to do this whole part where we calculate the public key and compare it with the contract_redeemscript. Consequently we also don't need the function `get_contract_privkey`.

GeneFerneau replied:
I replaced these tests with a function `SwapCoin::contract_type` that directly returns the type string:

- IncomingSwapCoin => "hashlock"
- OutgoingSwapCoin => "timelock"
- WatchOnlySwapCoin => "watchonly"

chris-belcher commented on src/maker_protocol.rs:
```diff
@@ -553,12 +553,20 @@ fn handle_sign_receivers_contract_tx(
+            if let Some(c) = wallet
                 .read()
                 .unwrap()
-                .find_swapcoin(&receivers_contract_tx_info.multisig_redeemscript)
-                .ok_or(Error::Protocol("multisig_redeemscript not found"))?
-                .sign_contract_tx_with_my_privkey(&receivers_contract_tx_info.contract_tx)?,
+                .find_incoming_swapcoin(&receivers_contract_tx_info.multisig_redeemscript)
+            {
+                c.sign_contract_tx_with_my_privkey(&receivers_contract_tx_info.contract_tx)?
+            } else {
+                wallet
+                    .read()
+                    .unwrap()
+                    .find_outgoing_swapcoin(&receivers_contract_tx_info.multisig_redeemscript)
+                    .ok_or(Error::Protocol("multisig_redeemscript not found"))?
+                    .sign_contract_tx_with_my_privkey(&receivers_contract_tx_info.contract_tx)?
+            },
```
Searching both kinds of swapcoins is wrong here. You should only search the list containing OutgoingSwapCoins (I think), because they are the only ones relevant for the SignReceiversContractTx.

chris-belcher commented:
I've done a first-pass review and written some comments. They're actually all related to one issue about not having three functions be in the SwapCoin trait.

There's also another issue about how in `maker_protocol.rs` its wrong to check both the `get_incoming_swapcoin` and `get_outgoing_swapcoin` functions, however that can be fixed later it doesn't have to be in this PR.

GeneFerneau commented:
> There's also another issue about how in maker_protocol.rs its wrong to check both the get_incoming_swapcoin and get_outgoing_swapcoin functions, however that can be fixed later it doesn't have to be in this PR.

I was wondering if there was only a need to check for `IncomingSwapCoin` or `OutgoingSwapCoin` in `handle_sign_receivers_contract_tx`, but didn't have enough context to make the call. So, defaulted to checking for both types, since that's what the previous functionality did.

I've added a fixup commit, and rebased on latest master. Let me know what you think. Will squash the fixup commit(s) once you're happy with the changes.

chris-belcher commented on src/wallet_sync.rs:
```diff
@@ -604,20 +786,25 @@ impl Wallet {
         &self,
         rpc: &Client,
-    ) -> Result<HashMap<[u8; 20], Vec<(ListUnspentResultEntry, &WalletSwapCoin)>>, Error> {
+    ) -> Result<HashMap<[u8; 20], Vec<(ListUnspentResultEntry, &dyn SwapCoin)>>, Error> {
```
Please change this whole function a little bit to return `HashMap::<[u8; 20]>, (Vec<(ListUnspentResultEntry, &IncomingSwapCoin)>, Vec<(ListUnspentResultEntry, &OutgoingSwapCoin)>)`. So make the IncomingSwapCoin and OutgoingSwapCoin be returned separately in different vectors.

chris-belcher commented:
Sorry about the delay on my end. Thanks for sticking with this.

Just one requested change now, and after that please squish the commits. I'll have a bit of time right now if you reckon you'll be delayed I can just merge this PR and do the requested change myself.

GeneFerneau commented:
> Sorry about the delay on my end. Thanks for sticking with this.

No worries, thanks for the review!

> Just one requested change now, and after that please squish the commits.

Will do

chris-belcher commented on src/main.rs:
```diff
@@ -173,52 +168,54 @@ fn display_wallet_balance(wallet_file_name: &PathBuf, long_form: Option<bool>) {
+            let balance: Amount = utxo_incoming_swapcoins
                 .iter()
-                .fold(Amount::ZERO, |acc, us| acc + us.0.amount);
+                .zip(utxo_outgoing_swapcoins.iter())
```
This assumes that the number of incoming swapcoins is the same as the number of outgoing swapcoins, which isn't always true (for example it wont be true if a coinswap has stalled, because then some peers will have outgoing swapcoins but no corresponding incoming swapcoin). I'd recommend using Iterator.chain() instead or something else similar.

GeneFerneau replied:
Fixed

chris-belcher commented:
I just tested if `wallet-balance` could show incomplete coinswaps. I started a maker and then did `coinswap-send` with the taker, then after the first message `waiting for funding transaction to confirm` I closed the maker and taker, and told regtest to generate a block (to confirm the funding transaction). But now running `wallet-balance` on the taker wallet doesn't show any incomplete coinswap but it should.

GeneFerneau commented:
> I just tested if `wallet-balance` could show incomplete coinswaps. I started a maker and then did `coinswap-send` with the taker, then after the first message `waiting for funding transaction to confirm` I closed the maker and taker, and told regtest to generate a block (to confirm the funding transaction). But now running `wallet-balance` on the taker wallet doesn't show any incomplete coinswap but it should.

Hopefully, the fix for the iterators in main, fix the reporting issue. If not, let me know, and I'll try to reproduce your test setup.

Sidenote: should I work on setting up automated integration tests for situations like the one you mentioned?

chris-belcher commented on src/wallet_sync.rs:
```diff
@@ -607,20 +789,50 @@ impl Wallet {
             .filter(|sc| sc.other_privkey.is_some())
             .map(|sc| read_hashvalue_from_contract(&sc.contract_redeemscript).unwrap())
             .collect::<HashSet<Hash160>>();
+        completed_coinswap_hashvalues.extend(
+            self.outgoing_swap_coins
+                .values()
+                .map(|sc| read_hashvalue_from_contract(&sc.contract_redeemscript).unwrap()),
+        );
```
Remove these lines. They add all the hashvalues of every outgoing swap coin regardless of whether it is incomplete or not. It results in outgoing swap coins never being returned by this function, which is the cause of the bug I found where incomplete swapcoins are not shown in `wallet-balance`.

GeneFerneau replied:
Thanks for retesting and debugging. Fixed

chris-belcher commented on src/wallet_sync.rs:
```diff
@@ -607,20 +789,50 @@ impl Wallet {
 
-        let completed_coinswap_hashvalues = self
-            .swap_coins
+        let mut completed_coinswap_hashvalues = self
```
```suggestion
        let completed_coinswap_hashvalues = self
```

chris-belcher commented on src/wallet_sync.rs:
```diff
@@ -94,56 +96,124 @@ const WATCH_ONLY_SWAPCOIN_LABEL: &str = "watchonly_swapcoin_label";
+    pub timelock_privkey: SecretKey,
+    pub funding_amount: u64,
+    pub others_contract_sig: Option<Signature>,
+    pub hash_preimage: Option<[u8; 32]>,
```
Nit: should use `Hash` rather than `[u8; 32]` here.

chris-belcher commented:
That last commit did not fix the issue. I thought it would be easier if I put print statements everywhere myself and I tracked down the cause, which I wrote in a review comment here just now.

Sidenote: Yes but not right now. I'm still working on code which will go alongside this PR which waits for the timeout branch of the contract transactions and then adds those coins to the wallet.
